From 46d45175ef74bf61e99c527f465b4d786f6124a1 Mon Sep 17 00:00:00 2001
From: Hui Peng <phui@google.com>
Date: Fri, 2 Dec 2022 02:02:01 +0000
Subject: [PATCH] Fix a use-after-free bug in
 AttributionProcessor::OnWakelockReleased

There is a use-after-free bug in AttributionProcessor::OnWakelockReleased
resulted from a well-known misuse of using iterators to delete
items in containers (the deleted items are used for calculating the next iterator
in the next round). This patch fix it with correct usage.

Note:
1. This is a cherry-pick of  If9f14d5fe2fbf2150f2ab0d1f90ce0f263399227
2. The regression test is: If40eb63e00c1a97e15dcdfdbbf12fad1070cd97b

Bug: 254774758
Ignore-AOSP-First: security
Test: atest bluetooth_test_gd_unit
Change-Id: I75576e59e0c81a82473a68a6c5ba3ce882a84f99
(cherry picked from commit 9774aeff84a834ae4403300b5ef88f0a4635e9ac)
Merged-In: I75576e59e0c81a82473a68a6c5ba3ce882a84f99
---
 gd/btaa/linux_generic/attribution_processor.cc | 11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/gd/btaa/linux_generic/attribution_processor.cc b/gd/btaa/linux_generic/attribution_processor.cc
index b38bb3a8b..4c0eac780 100644
--- a/gd/btaa/linux_generic/attribution_processor.cc
+++ b/gd/btaa/linux_generic/attribution_processor.cc
@@ -91,12 +91,15 @@ void AttributionProcessor::OnWakelockReleased(uint32_t duration_ms) {
     return;
   }
   // Trim down the transient entries in the aggregator to avoid that it overgrows
-  for (auto& it : btaa_aggregator_) {
+  auto it = btaa_aggregator_.begin();
+  while (it != btaa_aggregator_.end()) {
     auto elapsed_time_sec =
-        std::chrono::duration_cast<std::chrono::seconds>(cur_time - it.second.creation_time).count();
+        std::chrono::duration_cast<std::chrono::seconds>(cur_time - it->second.creation_time).count();
     if (elapsed_time_sec > kDurationTransientDeviceActivityEntrySecs &&
-        it.second.byte_count < kByteCountTransientDeviceActivityEntry) {
-      btaa_aggregator_.erase(it.first);
+        it->second.byte_count < kByteCountTransientDeviceActivityEntry) {
+      it = btaa_aggregator_.erase(it);
+    } else {
+      it++;
     }
   }
 }
-- 
2.39.1.456.gfc5497dd1b-goog

